tests/validator/validator.cpp

```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(errorMessageValidator, validator->issue(0)->description());
+}
+
+TEST(Validator, highIndexUnitsImport)
+{
+    const std::string errorMessageImporter =
+        "Import of units 'i_am_bad' from '"
+        + resourcePath("importer/") + "units_library.cellml' requires units named 'ps', which relies on child units named 'seconds', which cannot be found.";
+    const std::string errorMessageValidator =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Units reference 'seconds' in units 'ps' is not a valid reference to a local units or a standard unit type.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_units_that_have_unknown_children.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->errorCount());
+    EXPECT_EQ(errorMessageImporter, importer->error(0)->description());
+    EXPECT_EQ(model->units(1), importer->issue(0)->units());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessageValidator, validator->error(0)->description());
+}
+
+TEST(Validator, importInvalidUnitsNotDirectlyDeterminedFromImport)
+{
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps2' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Prefix 'plinco' of a unit referencing 'second' in units 'ps2' is not a valid integer or an SI prefix.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_units_that_are_invalid.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto unitsLibrary = parser->parseModel(fileContents("importer/units_library.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    validator->validateModel(unitsLibrary);
+    EXPECT_EQ(size_t(2), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage, validator->issue(0)->description());
+}
+
+TEST(Validator, importInvalidUnitsReference)
+{
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' refers to units 'ps3' which does not appear in 'units_library'.";
```

Comment:
We shouldn't mention the name of the imported model, but its file name (same in test `importInvalidComponentReference`):
```
Units 'i_am_bad' is imported from units 'ps3' which does not appear in 'units_library.cellml'.
```

Reply:
64cd2ed